Lab2 (#46)
UPB-FILS-MA/upb-fils-ma.github.io: docs/lab/02/02.md

:::

## Control GPIO Pins

As we saw in the previous chapters, the GPIO pins can be used to communicate with the environment through external elements. Next we will see how we can do this using Embassy-rs.

### Writing Rust code to control LEDs using GPIO pins.

The LEDs can be controlled by gpio because the pins can output electric current depending on their configuration (read the datasheet). Therefore, after you configure the pin to correspond with your led, it is necessary to control the voltage for the respective led, so we have to look at the panel of the GPIO pin and choose the appropriate resistance.

**Steps to turn on/off a led:**

1. First of all, we will define our LED and assign it the Level::Low property to be initially off.
```rust
let mut led = Output::new(p.PIN_17, Level::Low);
```

2. To control the flow of current passing through the pin, we have 2 functions to set the pin to low or high depending on the need.
```rust
led.set_high();
led.set_low();
```

> :::warning 
> 
> The GPIO pins output a standard output current depending on the microcontroller. To be sure that the LED will work normally and there is no risk of destruction, it is recommended to calculate and add a resistor to control the current level.
> 
> :::
> 
> <div align="center">
> ![LED GPIO](images/led-gpio17.png)
> </div>

Reviewer: ```suggestion
:::warning 

The maximum current that GPIO pins output depends on the microcontroller. To be sure that the LED will work normally and there is no risk of destruction, a resistor has to be added to limit the current level *below* the maximum GPIO output current.

<div align="center">
![LED GPIO](images/led-gpio17.png)
</div>

:::
```

I would use an image with the Raspberry Pi Pico, this one seems to be for the Raspberry Pi.

### Writing Rust code to read button state using GPIO pins.

> To read the input from the button, we must define the gpio we will use and also choose whether we will put a pull-up or pull-down resistor.

Reviewer: ```suggestion
To read the input from the button, we must define the GPIO we will use and also choose whether we will put a pull-up or pull-down resistor.
```

**Steps to turn read state of a button:**

1. We need to define the pin used and set the type of resistance used.
```rust
let button = Input::new(p.PIN_2, Pull::Up);
```

2. To read the input value we can use the is_high() function which will return true if the pin detects input current or false if the pin does not detect input current.
```rust
if button.is_high() {
    // Do something ...
} else {
    // Do something ...
}
```

<div align="center">
![Button GPIO](images/gpio-button.png)
</div>

## Advanced techniques for using GPIO pins

In this topic we will discuss several techniques used to solve some hardware problems that can appear using only software.

> ### Using async for efficient GPIO handling.

Reviewer: Please do this with a loop, reading continuously. We did not discuss async/await yet.

The loop runs infinitely, when using the GPIO pins, problems may arise because of this. We can block the loop until meeting certain signals coming on the GPIO.

```rust
let mut async_input = Input::new(p.PIN_16, Pull::Down);

loop {
  async_input.wait_for_high().await;
  info!("The button has been pressed");
}
```

In the case above, the loop is blocked until the button sends the high status.

### Debouncing techniques for stable input reading.

Noise is produced whenever a pushbutton or other switch is moved. Because the switch contact is made of metal and has some elasticity, there is some noise (contact). The switch literally bounces a few times once it makes contact with a metal surface when it is shifted into a new position. This contact is known as bounce. 

<div align="center">
![Button Bounce](images/button-bounce.png)
</div>

The image above shows the signal produced by a button when pressed.

> The most correct way to correct the bouncing problem is the hardware one, but there are also software methods to correct the problem. For more details and examples, consult the documentation from Embassy-rs and the examples provided by them.

Reviewer: I suggest adding a simple example here.

## Exercises

1. Create a Rust program using Embassy-rs set on high an LED connected to GPIO pin 0.
2. Connect an LED to the GP2 pin and create a program written in Embassy-rs to toggle the LED once every 2 seconds.

:::note

To put delay in the infinite loop, this is an example:

```rust
let delay = Duration::from_secs(1);
loop {
    // your code
    Timer::after(delay).await;
    // your code
}
```

> 

Reviewer: ```suggestion

For the purpose of this lab, please use `await` as is, think that for using the `Timer`, you have to add `.await` after the `after` function.

```

---
sidebar_position: 3
---

# 02 - GPIO

We will use the [Rust](https://www.rust-lang.org/) programming language and [Embassy-rs](https://embassy.dev/) for the labs.

> ## Definition and importance of GPIO in embedded systems

Reviewer: ```suggestion
## What is GPIO?
```

> GPIO, or General-Purpose Input/Output, is a crucial component in embedded systems, playing a fundamental role in connecting microcontrollers and microprocessors to the external world. GPIO refers to a set of pins on a microcontroller or microprocessor that can be individually configured to function as either input or output. These pins are used to interface with external devices, such as sensors, actuators, displays, and other components, allowing the embedded system to interact with its environment. Some GPIO pins may directly support standardized communication protocols like serial communication, SPI, I2C, PCM, and PWM. GPIO pins are of two types, analog or digital.

Reviewer: rephrase

</div>

### Digital Signal

A digital signal is a discrete representation of information using binary code, consisting of 0s and 1s. Unlike analog signals, digital signals are distinct, enabling precise and error-resistant communication. Digital signals are integral to modern computing, where data is processed using binary logic. They play a key role in telecommunications, ensuring reliable transmission through error correction. Digital audio formats, like MP3, encode sound as a series of binary digits for efficient storage and transmission. Digital signals are fundamental in modern displays, converting binary data into visual information on screens. They enable secure and robust storage in digital media such as hard drives and flash memory. Digital sensors, like those in cameras and touchscreens, convert physical phenomena into binary data for processing. The widespread use of digital signals is driven by their efficiency, accuracy, and adaptability in various technological applications.

<div align="center">
![Digital Signal](images/digitalsignal.png)
</div>

## Configuring GPIO Pins

GPIO pins can be used as outputs (LEDs, motors, buzzers) or as inputs (buttons, sensors).

> ### Basics of configuring GPIO pins for input and output.

Reviewer: use shorter sections and subsection names

> The GPIO pins are configured bare metal using the microcontroller's registers. Each pin has a separate register for configuring it as input or output and a separate register for writing or reading values from it. Below is a table with the memory addresses of the GPIO pins. We can see on the right a description of the address, and the offset for it. To write a register, we must compose a base address and an offset.

Reviewer: What is configured bare metal?

</div>

To see the number of a GPIO pin, we need the pinout scheme of microcontroller.

<div align="center">
![MCU Pins](images/mcuschematic.png)
</div>

:::info 

For a better understanding, please read [chapter 1.4.3](https://datasheets.raspberrypi.com/rp2040/rp2040-datasheet.pdf) of the datasheet.

:::

### Using the embassy-rs library for GPIO pin configuration

Because writing functionalities on an MCU is quite complicated and requires a lot of time using bare metal. Over time, several layers of abstraction were added to the programming languages and frameworks appeared to make the writing of new functionalities more efficient. One of these frameworks is Embassy-rs, which is a level of abstraction added on top of Rust to make the development process more efficient and faster.

**We have to follow several steps to configure a pin in embassy-rs:**

1. We need the specialized library from embassy rs for controlling the GPIO pins and certain components from it.

```rust
use gpio::{Input, Level, Output, Pull};
```

2. We have to define an gpio, choose its type and choose the corresponding pin number.

**Output**
```rust
let mut gpio_name = Output::new(p.PIN_n, Level::Low);
```

> You must replace n in PIN_n with the pin number. Level::Low sets the initial output of the pin, for example if you connect a led and initially want it to be off, set Level::Low, if you want it to be on, set Level::High.

Reviewer: ```suggestion
You must replace `n` in `PIN_n` with the pin number. `Level::Low` sets the initial output of the pin, for example if you connect a led and initially want it to be off, set Level::Low, if you want it to be on, set `Level::High`.
```

**Input**
```rust
let gpio_name = Input::new(p.PIN_n, Pull::Up);
```

> For the input type pins, as well as the output ones, we must set the pin number in PIN_n, n being the number in the pin diagram. Pull::Up or Pull::Down activates the pull up or pull down resistors of the GPIO pin.

Reviewer: ```suggestion
For the input type pins, as well as the output ones, we must set the pin number in `PIN_n`, `n` being the number in the pin diagram. `Pull::Up` or `Pull::Down` activates the internal pull up or pull down resistors of the GPIO pin.
```

```rust
led.set_high();
led.set_low();
```

:::warning 

The GPIO pins output a standard output current depending on the microcontroller. To be sure that the LED will work normally and there is no risk of destruction, it is recommended to calculate and add a resistor to control the current level.

:::

<div align="center">
![LED GPIO](images/led-gpio17.png)
</div>

### Writing Rust code to read button state using GPIO pins.

To read the input from the button, we must define the gpio we will use and also choose whether we will put a pull-up or pull-down resistor.

**Steps to turn read state of a button:**

1. We need to define the pin used and set the type of resistance used.
```rust
let button = Input::new(p.PIN_2, Pull::Up);
```

2. To read the input value we can use the is_high() function which will return true if the pin detects input current or false if the pin does not detect input current.
```rust
if button.is_high() {
    // Do something ...
} else {
    // Do something ...
}
```

<div align="center">
![Button GPIO](images/gpio-button.png)
</div>

> ## Analog Input/Output

Reviewer: This is not for this lab.

### Using async for efficient GPIO handling.

The loop runs infinitely, when using the GPIO pins, problems may arise because of this. We can block the loop until meeting certain signals coming on the GPIO.

```rust
let mut async_input = Input::new(p.PIN_16, Pull::Down);

loop {
  async_input.wait_for_high().await;
  info!("The button has been pressed");
}
```

In the case above, the loop is blocked until the button sends the high status.

### Debouncing techniques for stable input reading.

When a pushbutton or any switch's position is changed noise is generated. Some noise (contact) occurs because the switch contact is metal and it has elasticity. When the switch is moved to a new position it strikes a metal contact and physically bounces a few times. We call this contact bounce. 

<div align="center">
![Button Bounce](images/button-bounce.png)
</div>

The image above shows the signal produced by a button when pressed.

The correct method of correcting this problem is hardware, but it requires electronic expertise. A quick and much easier option is software correction. Using Embassy-rs will show as follows:

> ```rust

Reviewer: The example is too complicated.

    loop {
        // button pressed
        btn.debounce().await;
        let start = Instant::now();
        info!("Button Press");

        match with_deadline(start + Duration::from_secs(1), btn.debounce()).await {
            // Button Released < 1s
            Ok(_) => {
                info!("Button pressed for: {}ms", start.elapsed().as_millis());
                continue;
            }
            // button held for > 1s
            Err(_) => {
                info!("Button Held");
            }
        }

        match with_deadline(start + Duration::from_secs(5), btn.debounce()).await {
            // Button released <5s
            Ok(_) => {
                info!("Button pressed for: {}ms", start.elapsed().as_millis());
                continue;
            }
            // button held for > >5s
            Err(_) => {
                info!("Button Long Held");
            }
        }

        // wait for button release before handling another press
        btn.debounce().await;
        info!("Button pressed for: {}ms", start.elapsed().as_millis());
    }
}
```

## Exercises

1. Create a Rust program using embassy-rs set on high an LED connected to GPIO pin 0.

> 2. Connect an LED to the GP2 pin and create a program written in Embassy-RS to toggle the LED once every 2 seconds.

Reviewer: ```suggestion
2. Connect an LED to the GP2 pin and create a program written in Embassy-rs to toggle the LED once every 2 seconds.
```

Reviewer: We did not discuss timers with them. Please add an example of `delay`.

> 3. Connect a button to the GP3 pin and write a program in Embassy-RS that will write the message "The button was pressed" to the console every time the button is pressed.

Reviewer: ```suggestion
3. Connect a button to the GP3 pin and write a program in Embassy-rs that will write the message "The button was pressed" to the console every time the button is pressed.
```

We did no explain for to write debug messages.

---
sidebar_position: 3
description: Bare metal MMIO, PAC access and embassy-rs
---

# 02 - Memory Mapped IO & GPIO

The purpose of this lab is to understand how to start developing in [Rust](https://www.rust-lang.org/) for the
RP2040 MCU. The lab presents three examples:
 - **bare metal** development - writing directly to registers, actually writing a driver
 - **platform access crate** (PAC) - using an automatically generated crate from the MCUs SVD file, actually writing a driver, but with some kind of automation
 - **embassy-rs** - using the Rust standard [`embedded-hal`](https://docs.rs/embedded-hal/latest/embedded_hal/) implemented by the [Embassy-rs](https://embassy.dev/) framework.

:::info
The example of this lab is to blink an LED at a certain time interval.
:::

## Resources

1. **Raspberry Pi Ltd**, *[RP2040 Datasheet](https://datasheets.raspberrypi.com/rp2040/rp2040-datasheet.pdf)*
2. **Raspberry Pi Ltd**, *[Raspberry Pi Pico Datasheet](https://datasheets.raspberrypi.com/pico/pico-datasheet.pdf)*
3. **Raspberry Pi Ltd**, *[Raspberry Pi Pico W Datasheet](https://datasheets.raspberrypi.com/picow/pico-w-datasheet.pdf)*

## What is GPIO?

> General-Purpose Input/Output, or GPIO, is an essential part of embedded systems that serves as a vital conduit between microcontrollers and microprocessors and the outside world. A microcontroller or microprocessor's group of pins that can each be set to operate as an input or an output is referred to as GPIO. The purpose of these pins is to interface external components, including actuators, displays, sensors, and other devices, so that the embedded system may communicate with its surroundings. Standardized communication protocolsx like SPI, I2C, PCM, PWM, and serial communication may be directly supported by some GPIO pins. There are two varieties of GPIO pins: digital and analog.

Reviewer 2: ```suggestion
General-Purpose Input/Output, or GPIO, is an essential part of embedded systems that serves as a vital conduit between microcontrollers and microprocessors and the outside world. A microcontroller or microprocessor's group of pins that can each be set to operate as an input or an output is referred to as GPIO. The purpose of these pins is to interface external components, including actuators, displays, sensors, and other devices, so that the embedded system may communicate with its surroundings. Standardised communication protocols like SPI, I2C, PCM, PWM, and serial communication may be directly supported by some GPIO pins. There are two varieties of GPIO pins: digital and analog.
```

<div align="center">
![IO Bank0](images/gpio_mux.png)
</div>

## Bare metal

When using bare metal, developers interact directly with the hardware devices. They are responsible for all the
drivers and other components that they want to use. Mostly, bare meta development means *reading* and *writing* data
from and to the MCUs and peripheral registers.

The RP2040 (ARM Cortex-M0+) MCU uses *Memory Mapped Input Output* (MMIO). This means that the peripheral's registers
are mapped into the address space (in normal words, in the memory). Reading and writing data from and to these registers
is done by memory reads and writes.

Blinking an LED in bare metal programming means following these steps:

 1. Ask the rust compiler not to use the `std` library, as it depends on the operating system
 2. Write a `main` function and instruct the MCU to call it at startup (reset)
 3. Configure the Single Cycle IO (SIO) peripheral to set a pin as output
 4. Enable the *IO Bank0* peripheral of the RP2040
 5. Configure the *IO Bank0* peripheral so that it sets a certain pin as output
 6. Toggle the pin's value through the *SIO*'s registers
 7. Wait for an amount of time

>  8. Loop though steps 5 and 6

Reviewer 2: ```suggestion
 8. Loop through steps 5 and 6
```

the Rust compiler cannot rely on the `std` library. The two macros directives
ask the compiler not to link the `std` library and not to expect `main`
function.

```rust
#![no_main]
#![no_std]
```

:::warning

The bare metal code has to start with these two directives.

:::

### Bootloader

The RP2040 has a piece of software written in an internal ROM memory that is loaded 
when the MCU boots. This looks at the first 256 bytes of the Flash memory
to understand how it needs to start the MCU.

:::tip

While the ROM bootloader is rather small, its functionality is very similar to the PC's BIOS
boot sequence.

:::

The RP2040's datasheet explains the boot process. This is not very straight forward, and writing this
information requires a digital signature. The `rp2040-boot` crate provides the bootloader information
for booting with the `cortex-m-rt` crate.

Adding the following code to the Rust includes the bootloader.

```rust
#[link_section = ".boot_loader"]
#[used]
pub static BOOT_LOADER: [u8; 256] = rp2040_boot2::BOOT_LOADER_W25Q080;
```

### Entry

For this lab, we will use the `cortex-m-rt` crate. As the starting code for an MCU is usually processor

> and vendor dependant, this carte allows developers to get started a little faster. It implement the bare minimum

Reviewer 2: ```suggestion
and vendor dependent, this crate allows developers to get started a little faster. It implements the bare minimum
```

unsafe { write_volatile(gpio_ctrl, 5); }
```

:::info 

For a better understanding, please read subchapters 1.4.3, 2.1.12, 2.14, 2.19.1, 2.19.2, 2.19.15, 2.19.16 of the [datasheet](https://datasheets.raspberrypi.com/rp2040/rp2040-datasheet.pdf).

:::

### Configuring the Pad

The Pad peripheral is responsible for the electrical setup of the pins. It can configure the
the maximum output current, input pull up or pull down resistor.

The following table shows the *Pads* peripheral registers. Each GPIO pin has a corresponding register.

<div align="center">
![Pad Registers](images/gpio_pad_registers.png)
</div>

The following tables describe the `GPIOx` register. This register allows the configuration of several
electrical parameters.

<div align="center">
![Pad Pad Control 2](images/gpio_pad_ctrl_2.png)
</div>

<div align="center">
![Pad Pad Control 1](images/gpio_pad_ctrl_1.png)
</div>

Looking at the default values, when the MCU starts, pins are configured:

| | |
|-|-|
| Output Enabled | Yes |
| Input Enabled | Yes |
| Maximum Output Current | 4mA |
| Pull Input Resistor | Pull Down |
| Schmidt Trigger | Yes |
| Slew Rate | Slow |

> For this lab, the default value are fine, this peripheral can be considered as being properly setup.

Reviewer 2: ```suggestion
For this lab, the default values are fine, this peripheral can be considered as being properly setup.
```

### Wait for an amount of time

When using an operating system, developers usually have a function called `sleep` which asks the operating system
to suspend the process for a certain amount of time. In bare metal environment, with no framework or operating system,

> this is not available. We MCU timers can be used for this, the simplest (and most inefficient) of waiting is to

Reviewer 2: ```suggestion
this is not available. The MCU timers can be used for this, but the simplest (and most inefficient) of waiting is to
```

| [`embassy_rp`](https://docs.embassy.dev/embassy-rp/git/rp2040/index.html) | HAL Implementation |
| [`cortex-m`](https://docs.rs/cortex-m/latest/cortex_m/), [`cortex-m-rt`](https://docs.rs/cortex-m-rt/latest/cortex_m_rt/) | μ-architecture crates |
| [`rp_pac`](https://docs.embassy.dev/rp-pac/git/default/index.html) | Platform Access Crate |

:::info

The name *Embassy-rs* is derived form **Emb**edded **Asy**nchronous Rust.

:::

### Entry

Embassy-rs is a framework built on top of `cortex-m-rt` and provides its own method of defining
the entrypoint and bootloader.

```rust
use embassy_executor::Spawner;

#[embassy_executor::main]
async fn main(_spawner: Spawner) {
    let peripherals = embassy_rp::init(Default::default());
}
```

Instead of using the `entry` macro, embassy-rs provides the `embassy_executor::main` macro that
sets the `main` function as the startup point and defines the correct bootloader.

The `embassy_rp::init` function takes care of the peripheral initialization so that developers can jumo
right in and use them.

:::note

> Embassy-rs is designed to work in an asynchronous way and this is why the `main` function is defined as `async`. For te time being, just take it as a must and ignore it.

Reviewer 2: ```suggestion
Embassy-rs is designed to work in an asynchronous way and this is why the `main` function is defined as `async`. For the time being, just take it as a must and ignore it.
```

```rust
use gpio::{Level, Output};

// initialize PIN_n (replace n with a number) and set its
// default value to LOW (0)
let mut pin = Output::new(peripherals.PIN_n, Level::Low);

// set the pin value to HIGH (1)
pin.set_high();

// set the pin value to LOW (0)
pin.set_low();
```

:::tip

While the device initialization is specific to every hardware device (the example uses the 
`embassy_rp` crate that is for RP2040), the pin initialization and usage is portable. It
uses the same code, regardless of the MCU used.

:::

### Configure GPIO Input

Using a pin as input is very similar to using it as output.

```rust
use gpio::{Input, Pull};

let pin = Input::new(peripherals.PIN_n, Pull::Up);

if pin.is_high() {
    // Do something if the pin value is HIGH (1)
} else {
    // Do something if the pin value if LOW (0)
}
```

:::warning 

For a correct use of the buttons, use pull-up, pull-down resistors depending on the mode of operation of the button. Check [01 - Hardware Introduction](https://embedded-rust-101.wyliodrin.com/docs/lab/01#buttons)

:::

### Waiting

Embassy-rs provides support for suspending the execution of the software for an amount of time. It uses

> the [`Timer`](https://docs.rs/embassy-time/0.3.0/embassy_time/struct.Timer.html) structure form the

Reviewer 2: ```suggestion
the [`Timer`](https://docs.rs/embassy-time/0.3.0/embassy_time/struct.Timer.html) structure from the
```

<div align="center">
![IO Bank0](images/gpio_mux.png)
</div>

## Bare metal

When using bare metal, developers interact directly with the hardware devices. They are responsible for all the
drivers and other components that they want to use. Mostly, bare meta development means *reading* and *writing* data
from and to the MCUs and peripheral registers.

The RP2040 (ARM Cortex-M0+) MCU uses *Memory Mapped Input Output* (MMIO). This means that the peripheral's registers
are mapped into the address space (in normal words, in the memory). Reading and writing data from and to these registers
is done by memory reads and writes.

Blinking an LED in bare metal programming means following these steps:

 1. Ask the rust compiler not to use the `std` library, as it depends on the operating system
 2. Write a `main` function and instruct the MCU to call it at startup (reset)
 3. Configure the Single Cycle IO (SIO) peripheral to set a pin as output
 4. Enable the *IO Bank0* peripheral of the RP2040
 5. Configure the *IO Bank0* peripheral so that it sets a certain pin as output
 6. Toggle the pin's value through the *SIO*'s registers
 7. Wait for an amount of time

>  8. Loop though steps 5 and 6

Reviewer: Done.

### no-std

As the code written runs on an MCU without any framework or operating system,
the Rust compiler cannot rely on the `std` library. The two macros directives
ask the compiler not to link the `std` library and not to expect `main`
function.

```rust
#![no_main]
#![no_std]
```

:::warning

The bare metal code has to start with these two directives.

:::

### Bootloader

The RP2040 has a piece of software written in an internal ROM memory that is loaded 
when the MCU boots. This looks at the first 256 bytes of the Flash memory
to understand how it needs to start the MCU.

:::tip

While the ROM bootloader is rather small, its functionality is very similar to the PC's BIOS
boot sequence.

:::

The RP2040's datasheet explains the boot process. This is not very straight forward, and writing this
information requires a digital signature. The `rp2040-boot` crate provides the bootloader information
for booting with the `cortex-m-rt` crate.

Adding the following code to the Rust includes the bootloader.

```rust
#[link_section = ".boot_loader"]
#[used]

> pub static BOOT_LOADER: [u8; 256] = rp2040_boot2::BOOT_LOADER_W25Q080;

Reviewer: I think it should be enough to ask them to use this code and add the crate as a dependency. We can write a bare metal setup tutorial where we write this.

// use the CLR (+0x3000) trick that the RP2040
// provides, we have to read and write the
// `RESET` register.
// svd2rust does not know about this so
// it cannot generated the corresponding
// register
reset
    .reset()
    .modify(|r, w| unsafe { w.bits(r.bits() & !(1 << 5)) });
while reset.reset_done().read().bits() & (1 << 5) == 0 {}
```

#### Configure the GPIO

```rust
let io_bank0 = peripherals.IO_BANK0;

// write the value 5 to the FUNCSEL field
// Note: This actually writes the value 5 in FUNCSEL
//       and 0s in all the other field. While this
//       if fine for this example, usually a
//       read, modify, write sequence should be used
io_bank0
    .gpio(pin)
    .gpio_ctrl()
    .write(|w| unsafe { w.bits(5) });

// the correct way of doing this
io_bank0
    .gpio(pin)
    .gpio_ctrl()
    // the PAC crate provides fucntions for all the 
    // register's fields
    .modify(|_, w| unsafe { w.funcsel().bits(5) });
```

## Embedded HAL Implementation

`embassy-rs`

The bare metal and PAC require a lot of time and effort to develop applications.

The Rust [Embedded devices Working Group](https://www.rust-lang.org/governance/wgs/embedded) has designed 

> a  set of standard traits (interfaces) for interacting with an MCU. This is called the **Embedded Hardware Abstraction Layer**, or shortly Embedded HAL. The mai purpose is to define a common hardware interface that

Reviewer 3: ```suggestion
a  set of standard traits (interfaces) for interacting with an MCU. This is called the **Embedded Hardware Abstraction Layer**, or shortly Embedded HAL. The main purpose is to define a common hardware interface that
```

| [`embedded-hal`](https://docs.rs/embedded-hal/latest/embedded_hal/), [`embedded-hal-async`](https://docs.rs/embedded-hal-async/latest/embedded_hal_async/)| **Rust Embedded HAL (Standard)** |
| [`embassy_rp`](https://docs.embassy.dev/embassy-rp/git/rp2040/index.html) | HAL Implementation |
| [`cortex-m`](https://docs.rs/cortex-m/latest/cortex_m/), [`cortex-m-rt`](https://docs.rs/cortex-m-rt/latest/cortex_m_rt/) | μ-architecture crates |
| [`rp_pac`](https://docs.embassy.dev/rp-pac/git/default/index.html) | Platform Access Crate |

:::info

The name *Embassy-rs* is derived form **Emb**edded **Asy**nchronous Rust.

:::

### Entry

Embassy-rs is a framework built on top of `cortex-m-rt` and provides its own method of defining
the entrypoint and bootloader.

```rust
use embassy_executor::Spawner;

#[embassy_executor::main]
async fn main(_spawner: Spawner) {
    let peripherals = embassy_rp::init(Default::default());
}
```

Instead of using the `entry` macro, embassy-rs provides the `embassy_executor::main` macro that
sets the `main` function as the startup point and defines the correct bootloader.

> The `embassy_rp::init` function takes care of the peripheral initialization so that developers can jumo

Reviewer 3: ```suggestion
The `embassy_rp::init` function takes care of the peripheral initialization so that developers can jump
```
